lib/utils/fanoutbuffer/buffer.go

```diff
@@ -0,0 +1,431 @@
+	}
+}
+
+// Close permanently closes the fanout buffer. Note that all cursors are terminated immediately
+// and may therefore miss items that had not been read at the time Close is called.
+func (b *Buffer[T]) Close() {
+	b.write(func() {
+		b.closed = true
+		close(b.notify)
+	})
+}
+
+// NewCursor gets a new cursor into the buffer. Cursors *must* be closed as soon as they are
+// no longer being read from. Stale cursors may cause performance degredation.
+func (b *Buffer[T]) NewCursor() *Cursor[T] {
+	var c uint64
+	b.write(func() {
+		if b.closed {
+			return
+		}
+		c = b.pos + b.len
+		b.cursors++
+	})
+
+	cursor := &Cursor[T]{
+		buf:    b,
+		cursor: c,
+	}
+	runtime.SetFinalizer(cursor, finalizeCursor[T])
+	return cursor
+}
+
+// Append appends to the buffer and wakes all dormant cursors.
+func (b *Buffer[T]) Append(items ...T) {
+	var notify chan struct{}
+
+	b.write(func() {
+		if b.closed {
+			panic("Append called on closed fanout buffer")
+		}
+
+		// free up slots that are no longer needed
+		b.cleanupSlots()
+
+		for _, item := range items {
+			// ensure that there is at least one free slot in ring
+			b.ensureSlot()
+
+			// insert item into next free slot in ring
+			idx := int((b.pos + b.len) % b.cfg.Capacity)
+			b.ring[idx].item = item
+			b.ring[idx].wait.Store(b.cursors)
+			b.len++
+		}
+
+		notify = b.notify
+		b.notify = make(chan struct{})
+	})
+
+	// notify all waiting cursors. we do this outside of the
+	// lock to (theoretically) reduce potential contention.
+	close(notify)
+}
+
+// cleanupSlots frees up enries that have been seen by all cursors and clears the backlog if
+// it has exceeded its grace period.
+func (b *Buffer[T]) cleanupSlots() {
+	// trim items from overflow that have been seen by all cursors or are past their expiry
+	now := b.cfg.Clock.Now()
+	var clearOverflowTo int
+	for i := 0; i < len(b.overflow); i++ {
+		clearOverflowTo = i
+		if b.overflow[i].wait.Load() > 0 && b.overflow[i].expires.After(now) {
+			break
+		}
+	}
+
+	clear(b.overflow[:clearOverflowTo])
+	b.overflow = b.overflow[clearOverflowTo:]
+
+	// clear overflow start state if overflow is empty
+	if len(b.overflow) == 0 {
+		b.overflow = nil
+	}
+
+	// clear items from ring that have been seen by all cursors
+	for b.len > 0 && b.ring[int(b.pos%b.cfg.Capacity)].wait.Load() == 0 {
+		b.ring[int(b.pos%b.cfg.Capacity)] = entry[T]{}
+		b.pos++
+		b.len--
+	}
+}
+
+// ensure that there is at least one open slot in the ring. overflows items as needed.
+func (b *Buffer[T]) ensureSlot() {
+	if b.len < b.cfg.Capacity {
+		// we have at least one free slot in ring, no need for overflow
+		return
+	}
+
+	// copy oldest entry to overflow. note the that it would actually be safe to just append the
+	// entire `entry` value rather than this append -> load -> store pattern that we follow
+	// here, but 'go vet' doesn't understand our locking semantics so we need to avoid implicitly
+	// copying the atomic, even when the write lock is held.
```

Comment:
Not sure if this comment is still entirely relevant, but it doesn't hurt to keep it either, up to you.
```suggestion
	// copy oldest entry to overflow.
```